### ISSUE TITLE
travis builds

### ISSUE THREAD

**Issue by stnava** [stnava/ANTsR#10]
@armaneshaghi - looks like we need several different options for travis builds

* i'd like to be able to automate releases from travis

* would also like to see RMD check results ( has to be done separately from above )

* coverage numbers would be nice - but this seems to time out 

seems that , b/c of limited travis time, we can't do all at once ...

any ideas about how to approach this?

**Comment by armaneshaghi** [comment]
Brian, 

I'm working on coverage numbers, there is still an issue I have opened on covr package, still waiting for their response. 

To speed up travis build and find time for `R CMD check`s  we need to separate ITK from ANTsR. I have been thinking about this and it is possible, if we automatically build ITK repo and upload it somewhere  (independent of ANTsR). This could be automated to minimise maintenance. 

We can automate releases from travis, we can deploy Amazon's S3, for example. But let's postpone this when we have resolved the above issues, which I think are more essential now. 

I'm now working on coverage numbers + writing test codes.

**Comment by armaneshaghi** [comment]
https://github.com/jimhester/covr/issues/55

**Comment by stnava** [comment]
good to know - @jeffduda  was also thinking about an itkR package which would also help ... such packages dont need to be very complex and may be low maintenance as well.

**Comment by jeffduda** [comment]
This is a fairly minimal combo of ITK and R. Works on OSX but working on
some linking issues in linux.
https://github.com/jeffduda/itkBSplineR

**Comment by armaneshaghi** [comment]
thanks this is something we need to do after getting a proper build with travis and coveralls, do you have the make file that could work with this inside ANTs(R)?

**Comment by stnava** [comment]
main thing that needs to be done is the libraries get packaged / shipped with the installed R package ... then we link to them via something along the lines of  what's done here:

https://github.com/stnava/ANTsR/blob/master/configure#L4-L7

so this itkr package would have to store / return whatever header / library files are needed for compilation.

**Comment by stnava** [comment]
similarly here:

https://github.com/stnava/ANTsR/blob/master/src/Makevars#L4-L12

**Comment by jeffduda** [comment]
Right, my configure and Makevars are basically the same as in ANTsR. If I
point the Makevars to the ANTsR compiled ITK libraries, then it works fine.
So the issue seems to be getting ITK compiled with the correct
settings/flags.

**Comment by bkandel** [comment]
From Travis docs, it sounds like there is no time limit for builds, only
for test suites:
http://docs.travis-ci.com/user/build-configuration/#Job-Timeouts  Is that
information incorrect or misleading?

I'd like to avoid splitting ANTsR into a different RITK package if
possible--just one more thing to maintain and keep track of.

**Comment by stnava** [comment]
it may be a bit more complicated ... i.e. figuring out how to get the *built and installed itkR package* to properly keep the itk library/header files w/o violating CRAN policies .... then these functions (akin to Rcpp::: ) would just return these locations  ....  if the package is not on cran, then it's not going to be all that useful ...

**Comment by stnava** [comment]
@bkandel  ultimately we might just keep the same structure we have, assuming we can get over these last 2 hurdles  ... https://github.com/stnava/ANTsR/issues/8#issuecomment-73693477 ...  but i think jeff's itkr package would still be useful if implemented to completion.

**Comment by bkandel** [comment]
Right, the package would definitely have to be on cran or we won't be able
to use it at all.  The only way I see this could help things is if 1) we
get the RITK package on cran 2) installing that via binary package install
from cran is faster than building it from source in antsr 3) The package
exports everything we need so that we can use the necessary functions
without anything illegal, like ':::' calls.  So that's why I'm skeptical
that it's worth the effort.

**Comment by jefferis** [comment]
Great that you chaps are moving to an itkr style package - should be good for eg travis builds. 

I'm not sure if you've seen, but Rcpp has a mechanism for getting the path to libraries distributed by another Rcpp-based package. If you can't find the place in the rcpp docs I'll see if I can dig it out. Best, Greg.

**Comment by stnava** [comment]
greg - that Rcpp pointer would be helpful ...   our travis builds are progressing w/help from @armaneshaghi ... was able to successfully deploy binaries from travis though i've not tested the linux one yet:   https://github.com/stnava/ANTsR/releases

**Comment by bkandel** [comment]
When I run R CMD INSTALL ANTsR --clean, it takes care of those .git files.
I wonder if there is something in the configuration of the package build
that is skipping the clean step.

**Comment by stnava** [comment]
yes - possible that there is some unright about the configuration / clean /
ignore setup  ... maybe a "truly clean" build would yield different results
too e.g.

git clone ANTsR

R CMD build ANTsR #  to create a package

R CMD check ANTsR.tar.gz #  to check a package

R CMD INSTALL ANTsR.tar.gz #  to install a (source) package




brian

**Comment by bkandel** [comment]
I'm still mystified by this.  When I build using R CMD INSTALL ANTsR
--clean, these files go away, but when I do R CMD CHECK, I still get the
same errors.

Also, just as a note for testing, we need to use R-devel:
http://cran.r-project.org/web/packages/policies.html#Submission

**Comment by stnava** [comment]
good point re: R-devel .... am not sure about the damn git files either ...
we can (if need be) make a cran branch and do some brute force rm w/in code
in that branch alone if need be


brian

**Comment by bkandel** [comment]
I couldn't tell easily--are the rest of the files to be deleted (*.so etc.)
being deleted correctly?

**Comment by armaneshaghi** [comment]
I spent quite a while on finding a way to include coverage testing.It seems that at some point the coverage testing package calls devtools::load_all() that wants to compile package, that for now fails for ANTsR (not sure why, perhaps because of some Rcpp conventions we're not following). It would be almost impossible to build the package only once during one travis session, because we need to run the compiler without optimization, after building ANTsR, so that we can count the number of lines being tested, which is going to take a long time for travis (50 minutes limit).

**Comment by stnava** [comment]
yes - i think so ... or being ignored

i believe the process is:

R CMD build => just packs up the source
R CMD check => unpacks, builds anew does checking etc ...

so the build directory will ignore stuff that's in the source that we dont
want to be "packed up"

then the check creates new "junk" (during compilation process) that should
( i thought ) be deleted by cleanup - but the cleanup seems to somehow miss
these files ( or rbuildignore is erroneously missing them )

we could investigate this by explicitly looking at what's in the tar.gz ...


brian

**Comment by bkandel** [comment]
Answer may be here:
https://stat.ethz.ch/pipermail/r-devel/2007-March/044987.html
.Rbuildignore is not ".Rcheckignore", and checks do not pay attention to
it.
Not sure how "cleanup" fits into all of this (if it's called and when
etc.).

**Comment by jefferis** [comment]
First, a disclaimer, I am no expert although I have several packages on CRAN including one based on Rcpp. 

Anyway, re your configure/clean setup, my understanding is that:

1. **configure** should not call make.
   It should set things up so that everything is ready for the package build process, which will then call make.
   see e.g. http://cran.r-project.org/doc/manuals/r-release/R-exts.html#Configure-and-cleanup

2. If anything removes your Makefile, it should be **cleanup**

AFAICS this is why devtools::load_all() currently fails to build ANTsR. Commenting out the last two lines gives me successful devtools::load_all().

Best,

Greg.

**Comment by armaneshaghi** [comment]
thanks for the comment @jefferis  this is great, have been scratching my head with load_all() for a day or two :-)

**Comment by armaneshaghi** [comment]
https://github.com/cran/RODBC/blob/master/configure

**Comment by stnava** [comment]
@bkandel suggested that we push to a 2nd antsr repo ( a decoy ) that would be used just to travis run R CMD check in a timely fashion .... we may try this.

**Comment by armaneshaghi** [comment]
nice, a new level of maintenance with a new repo, but I guess there is no
other option for now.

**Comment by bkandel** [comment]
Should not add any additional maintenance.  We'd have hooks that do all
this automatically.  I.e. if we forget about this we just ignore it and it
runs happily for eternity.

2015-03-12 13:24 GMT-04:00 Arman Eshaghi <notifications@github.com>:

> nice, a new level of maintenance with a new repo, but I guess there is no
> other option for now.
>

**Comment by armaneshaghi** [comment]
do you know how this can be done? I like ITK's kwrobot that automatically
pushes to github, but I guess we are not going to assign a server to do
this.

**Comment by bkandel** [comment]
I thought that you can put hooks on your github repo that whenever the repo
gets a push, it pushes the change to a different repo.  That requires no
manual intervention.  We can alternatively add a remote repository and then
automatically push there.

2015-03-12 13:39 GMT-04:00 Arman Eshaghi <notifications@github.com>:

> do you know how this can be done? I like ITK's kwrobot that automatically
> pushes to github, but I guess we are not going to assign a server to do
> this.
>

**Comment by armaneshaghi** [comment]
yes, seems that git already has done this
http://stackoverflow.com/questions/3669636/git-automatically-keep-a-secondary-repo-in-sync-with-a-primary-repo

**Comment by bkandel** [comment]
I'd avoid a cron job running on our own server if at all possible.

2015-03-12 13:48 GMT-04:00 Arman Eshaghi <notifications@github.com>:

> yes, seems that git already has done this
>
> http://stackoverflow.com/questions/3669636/git-automatically-keep-a-secondary-repo-in-sync-with-a-primary-repo
>
>

**Comment by stnava** [comment]
just to be clear: this will not require additional maintenance.  and it will be "opt-in" for developers.  but recommended for anyone contributing regularly.  will basically amount to:

git add remote antsrCranCheck ... push 

and not much more.

**Comment by stnava** [comment]
this issue superseded by a new one
